ERCS/erc-7641.md

```diff
@@ -0,0 +1,189 @@
+     * @param account The address of the token holder
+     * @param snapshotId The snapshot id
+     * @param token The address of the revenue token
+     * @return The amount of revenue token claimable
+     */
+    function claimableERC20(address account, uint256 snapshotId, address token) external view returns (uint256);
+
+    /**
+     * @dev A function to calculate the amount of ERC-20 redeemable by a token holder upon burn
+     * @param amount The amount of token to burn
+     * @param token The address of the revenue token
+     * @return The amount of revenue token redeemable
+     */
+    function redeemableERC20OnBurn(uint256 amount, address token) external view returns (uint256);
+}
+```
+
+## Rationale
+
+### Revenue Sharing Mechanism
+
+We implement a revenue sharing mechanism wherein any token holder can claim a proportional share from the revenue pool. To ensure regular and transparent revenue distribution, we have incorporated the snapshot method, capturing both the token balance and the associated claimable revenue token balance. Periodic invocation of the snapshot method, corresponding to distinct revenue-sharing processes, is required. During each snapshot, token holders are empowered to claim a proportionate share from the revenue pool, creating a systematic and equitable distribution mechanism for participants.
+
+### `snapshot` interface
+
+We specify a `snapshot` interface to snapshot the token balance and the claimable revenue token balance. This functionality ensures correctness in tracking token holdings, facilitating a transparent record of each token portfolio. Regular invocation of the snapshot function is essential to maintain up-to-date records. The `snapshot` interface returns a unique `snapshotId`, allowing access to the corresponding token balance and claimable revenue token balance associated with that specific snapshot. This systematic approach enhances the correctness and reliability of historical data retrieval, providing users with comprehensive insights into their token and revenue token balances at different points in time.
+
+### `claimableRevenue` interface
+
+We specify a `claimableRevenue` interface to calculate the amount of ETH claimable by a token holder at a certain snapshot. We will share the revenue between two consecutive snapshots. Specifically, assuming that the revenue between two snapshots is `R`, we will specify a revenue sharing ratio `p`, ranging from 0%-100%, and we will share the revenue of `pR` to different token holders according to the token ratio. Specifically,  the amount of ETH claimable by a token holder with `amount` tokens at a certain snapshot is `pR * amount / totalAmount` , where `totalAmount`  denotes the total amount of [ERC-7641](./eip-7641.md) token. Noted that the remaining revenue of `(1-p)R` will be retained in the revenue pool, and we can take out this part of revenue through burning.
```

Comment:
This paragraph sounds like it's creating requirements on implementations? For example, "the amount of ETH claimable [...] is `pR * amount / totalAmount`" seems like an instruction to implementers.

If these are requirements, they'll need to be moved to the Specification section. The _reasoning_ behind why that formula exists should remain in the Rationale section.

The idea is that the Specification section, taken alone, should be enough for someone to write a compliant implementation of this proposal. The other sections (Rationale included) are here for context and history purposes.

Reply:
It's more of a recommended/reference implementation. How do you recommend us to change the wording here to reflect this?

Reply:
We have changed the wording to reflect that it is an example. Please review!